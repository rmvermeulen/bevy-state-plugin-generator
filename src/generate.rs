#[cfg(test)]
mod tests;

use indoc::formatdoc;
use itertools::Itertools;
use nom::AsChar;

use crate::PluginConfig;
use crate::parsing::parse_states_text;
use crate::processing::{ProcessingError, try_parse_node_into_final_source};

pub(super) const REQUIRED_DERIVES: &[&str] =
    &["Hash", "Default", "Debug", "Clone", "PartialEq", "Eq"];

pub fn get_package_info() -> String {
    let pkg = env!("CARGO_PKG_NAME");
    #[cfg(not(test))]
    let version = env!("CARGO_PKG_VERSION");
    #[cfg(test)]
    let version = "[CARGO_PKG_VERSION]";
    format!("{pkg} v{version}")
}

pub fn generate_debug_info(src_path: &str, source: &str) -> String {
    let lines = source.lines().map(|line| format!("// {line}")).join("\n");
    let pkg_info = get_package_info();
    formatdoc! {"
        // generated by {pkg_info}
        // src: {src_path}
        {lines}
    "}
}

#[cfg(feature = "rustfmt")]
fn try_format_source(source: &str) -> std::io::Result<String> {
    duct::cmd!("rustfmt")
        .stdin_bytes(source)
        .stderr_to_stdout()
        .read()
}

pub(crate) fn format_source<S: AsRef<str>>(source: S) -> String {
    let source = source.as_ref();
    #[cfg(feature = "rustfmt")]
    let source = try_format_source(source).unwrap_or_else(|_| source.to_owned());
    #[cfg(not(feature = "rustfmt"))]
    let source = source.to_owned();

    if source.ends_with(|c: char| c.is_newline()) {
        source
    } else {
        source + "\n"
    }
}

pub fn generate_state_plugin_source(
    input_source: &str,
    plugin_config: PluginConfig,
    src_path: Option<&str>,
) -> Result<String, ProcessingError> {
    let parse_nodes = parse_states_text(input_source)?;
    let mut output = parse_nodes
        .into_iter()
        .map(|parse_node| try_parse_node_into_final_source(parse_node, plugin_config.clone()))
        .collect::<Result<Vec<_>, _>>()?
        .join("\n");
    if let Some(src_path) = src_path {
        let debug_info = generate_debug_info(src_path, input_source);
        output = [debug_info, output].join("\n");
    }
    Ok(format_source(output))
}
