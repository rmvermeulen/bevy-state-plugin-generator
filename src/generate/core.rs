use indoc::formatdoc;
use itertools::Itertools;
use nom::AsChar;

use crate::parsing::parse_config;
use crate::prelude::PluginConfig;
use crate::processing::{ProcessingError, convert_nodes_into_plugin_source};

pub(crate) const REQUIRED_DERIVES: &[&str] =
    &["Hash", "Default", "Debug", "Clone", "PartialEq", "Eq"];

pub(crate) fn get_package_info() -> String {
    let pkg = env!("CARGO_PKG_NAME");
    #[cfg(not(test))]
    let version = env!("CARGO_PKG_VERSION");
    #[cfg(test)]
    let version = "[CARGO_PKG_VERSION]";
    format!("{pkg} v{version}")
}

pub(crate) fn generate_debug_info(src_path: &str, source: &str) -> String {
    let lines = source.lines().map(|line| format!("// {line}")).join("\n");
    let pkg_info = get_package_info();
    formatdoc! {"
        // generated by {pkg_info}
        // src: {src_path}
        {lines}
    "}
}

pub(crate) fn format_source<S: AsRef<str>>(source: S) -> Result<String, String> {
    duct::cmd!("rustfmt")
        .stdin_bytes(source.as_ref())
        .stderr_to_stdout()
        .read()
        .map_err(|e| format!("rustfmt: {e:?}"))
        .map(|source| {
            if source.ends_with(|c: char| c.is_newline()) {
                source
            } else {
                source + "\n"
            }
        })
}

pub(crate) fn generate_state_plugin_source(
    input_source: &str,
    plugin_config: PluginConfig,
    src_path: Option<&str>,
) -> Result<String, ProcessingError> {
    let (unparsed, nodes) = parse_config(input_source)?;
    let mut output = convert_nodes_into_plugin_source(nodes, plugin_config)?;

    // if we're writing to a file we add a header with some information
    output = if let Some(src_path) = src_path {
        if !unparsed.trim().is_empty() {
            return Err(ProcessingError::InvalidConfig(unparsed.to_string()));
        }
        let debug_info = generate_debug_info(src_path, input_source);
        [debug_info, output].join("\n")
    } else {
        [unparsed, &output].join("\n")
    };

    format_source(&output).map_err(ProcessingError::Formatting)
}
