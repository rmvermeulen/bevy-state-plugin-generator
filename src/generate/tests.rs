use std::{convert::identity, time::Duration};

use super::*;
use insta::{assert_debug_snapshot, assert_snapshot};
use rstest::{fixture, rstest};

use crate::set_snapshot_suffix;
use crate::{NamingScheme, PluginConfig};

#[rstest]
#[timeout(Duration::from_millis(250))]
async fn test_format_source() {
    let formatted = format_source("fn main(){println!(\"Hello, world!\");}");
    assert_snapshot!(formatted, @r#"
        fn main() {
            println!("Hello, world!");
        }
        "#);
}

#[rstest]
fn test_generate_states_plugin() {
    let states = StateNode::enumeration("GameState", [
        StateNode::singleton("Loading"),
        StateNode::enumeration("Ready", [
            StateNode::enumeration("Menu", [
                StateNode::singleton("Main"),
                StateNode::singleton("Options"),
            ]),
            StateNode::enumeration("Game", [
                StateNode::singleton("Playing"),
                StateNode::singleton("Paused"),
                StateNode::singleton("GameOver"),
            ]),
        ]),
    ]);
    assert_snapshot!(generate_plugin_source(Rc::new(states), Default::default()));
}

#[rstest]
#[case("root.txt", "RootState")]
#[case("fruits.txt", "Apple Orange { O1 O2 }")]
fn test_generate_debug_info(#[case] src_path: &str, #[case] source: &str) {
    set_snapshot_suffix!("{src_path}");
    assert_snapshot!(generate_debug_info(src_path, source));
}

fn test_plugin_formatted(root_node: Rc<StateNode>, plugin_config: PluginConfig) -> String {
    format_source(generate_plugin_source(root_node, plugin_config))
}

#[rstest]
#[cfg_attr(
    feature = "comments",
    case("comments.txt", Rc::new(StateNode::enum_empty("GameState")),)
)]
#[case("simple.txt", Rc::new(StateNode::enumeration("GameState", [
    StateNode::singleton("Loading"),
    StateNode::enumeration("Ready", [
        StateNode::singleton("Menu"),
        StateNode::singleton("Game"),
    ]),
])))]
#[case("fruits.txt", Rc::new(StateNode::enumeration("GameState", [
    StateNode::singleton("Loading"),
    StateNode::enumeration("Ready", [
        StateNode::enumeration("Menu", [
            StateNode::singleton("Main"),
            StateNode::singleton("Options"),
        ]),
        StateNode::enumeration("Game", [
            StateNode::singleton("Playing"),
            StateNode::singleton("Paused"),
            StateNode::singleton("GameOver"),
        ]),
    ]),
])))]
fn test_generate_plugin_source(#[case] src_path: &str, #[case] root_node: Rc<StateNode>) {
    set_snapshot_suffix!("{src_path}");
    assert_snapshot!(test_plugin_formatted(root_node, Default::default()));
}

#[cfg(feature = "lists")]
#[rstest]
#[case("root.txt", "RootState", PluginConfig::default())]
#[case(
    "mixed-nested-states.txt",
    "A { B [C[ } D { E F [ G H ] I }",
    PluginConfig::default()
)]
fn test_generate_full_source(
    #[case] src_path: &str,
    #[case] source: &str,
    #[case] plugin_config: PluginConfig,
) {
    set_snapshot_suffix!("{src_path}");
    assert_snapshot!(
        generate_state_plugin_source(src_path, source, plugin_config).unwrap_or_else(identity)
    );
}

#[rstest]
#[case("root.txt", "RootState", NamingScheme::Full)]
#[case("root.txt", "RootState", NamingScheme::Short)]
fn test_naming_scheme(#[case] src_path: &str, #[case] source: &str, #[case] scheme: NamingScheme) {
    set_snapshot_suffix!("{src_path}_{scheme:?}");
    assert_snapshot!(
        generate_state_plugin_source(src_path, source, PluginConfig {
            scheme,
            ..Default::default()
        })
        .unwrap_or_else(identity)
    );
}

#[fixture]
fn root_source_state() -> SourceState {
    SourceState {
        name: "Game".to_string(),
        variant: "Menu".to_string(),
    }
}
#[fixture]
fn nested_node() -> StateNode {
    StateNode::enumeration("Menu", [
        StateNode::singleton("Main"),
        StateNode::enumeration("Options", [
            StateNode::singleton("Graphics"),
            StateNode::singleton("Audio"),
            StateNode::singleton("Gameplay"),
        ]),
        StateNode::enumeration("Continue", [
            StateNode::singleton("Save"),
            StateNode::singleton("Load"),
        ]),
    ])
}

#[rstest]
fn test_generate_all_type_definitions_full(
    #[from(root_source_state)] source: SourceState,
    #[from(nested_node)] node: StateNode,
) {
    let typenames = generate_all_type_definitions(&node, Context {
        source_state: Some(source.clone()),
        naming_scheme: NamingScheme::Full,
        ..Default::default()
    })
    .take()
    .into_iter()
    .map(|td| td.typename)
    .collect_vec();
    assert_debug_snapshot!(typenames, @r#"
    [
        "GameMenu",
        "GameMenuMain",
        "GameMenuOptions",
        "GameMenuOptionsGraphics",
        "GameMenuOptionsAudio",
        "GameMenuOptionsGameplay",
        "GameMenuContinue",
        "GameMenuContinueSave",
        "GameMenuContinueLoad",
    ]
    "#);
}

#[rstest]
fn test_generate_all_type_definitions_shortened(
    #[from(root_source_state)] source: SourceState,
    #[from(nested_node)] node: StateNode,
) {
    assert_debug_snapshot!(
        generate_all_type_definitions(&node, (source, NamingScheme::Short).into())
            .take().into_iter().map(|td| td.typename).collect_vec(),
        @r#"
    [
        "GameMenu",
        "MenuMain",
        "MenuOptions",
        "OptionsGraphics",
        "OptionsAudio",
        "OptionsGameplay",
        "MenuContinue",
        "ContinueSave",
        "ContinueLoad",
    ]
    "#);
}

#[rstest]
fn test_generate_all_type_definitions_none(
    #[from(root_source_state)] source: SourceState,
    #[from(nested_node)] node: StateNode,
) {
    assert_debug_snapshot!(
        generate_all_type_definitions(&node, (source, NamingScheme::None).into())
            .take().into_iter().map(|td| td.typename).collect_vec(),
        @r#"
    [
        "Menu",
        "Main",
        "Options",
        "Graphics",
        "Audio",
        "Gameplay",
        "Continue",
        "Save",
        "Load",
    ]
    "#);
}

#[rstest]
fn snapshots(
    #[values(NamingScheme::Full, NamingScheme::Short)] scheme: NamingScheme,
    #[from(root_source_state)] source: SourceState,
    #[from(nested_node)] node: StateNode,
) {
    set_snapshot_suffix!("{scheme:?}");
    assert_snapshot!(generate_all_type_definitions(
        &node,
        (source, scheme).into()
    ));
}

#[rstest]
fn snapshot1() {
    assert_snapshot!(generate_all_type_definitions(
        &StateNode::singleton("Alpha"),
        SourceState {
            name: "GameState".to_string(),
            variant: "Alpha".to_string()
        }.into()
    ), @r"
    #[derive(bevy::prelude::SubStates, Hash, Default, Debug, Clone, PartialEq, Eq)]
    #[source(GameState = GameState::Alpha)]
    pub struct Alpha;
    ");
}

#[rstest]
fn snapshot1a() {
    assert_snapshot!(generate_all_type_definitions( &StateNode::singleton("Alpha"), NamingScheme::Full.into()), @r"
        #[derive(bevy::prelude::States, Hash, Default, Debug, Clone, PartialEq, Eq)]
        pub struct Alpha;
        ");
}

#[rstest]
fn snapshot2() {
    assert_snapshot!(generate_all_type_definitions(
        &StateNode::enumeration("Alpha", [StateNode::singleton("Beta")]),
        (SourceState {
            name: "GameState".to_string(),
            variant: "Alpha".to_string()
        },
        NamingScheme::Full).into()
    ), @r"
        #[derive(bevy::prelude::SubStates, Hash, Default, Debug, Clone, PartialEq, Eq)]
        #[source(GameState = GameState::Alpha)]
        pub enum GameStateAlpha { #[default] Beta }

        #[derive(bevy::prelude::SubStates, Hash, Default, Debug, Clone, PartialEq, Eq)]
        #[source(GameStateAlpha = GameStateAlpha::Beta)]
        pub struct GameStateAlphaBeta;
        ");
}

#[rstest]
fn snapshot2a() {
    assert_snapshot!(generate_all_type_definitions(
        &StateNode::enumeration("Alpha", [StateNode::singleton("Beta")]),
        NamingScheme::Full.into()
    ), @r"
        #[derive(bevy::prelude::States, Hash, Default, Debug, Clone, PartialEq, Eq)]
        pub enum Alpha { #[default] Beta }

        #[derive(bevy::prelude::SubStates, Hash, Default, Debug, Clone, PartialEq, Eq)]
        #[source(Alpha = Alpha::Beta)]
        pub struct AlphaBeta;
        ");
}

#[cfg(feature = "lists")]
#[rstest]
fn snapshot3() {
    assert_snapshot!(generate_all_type_definitions(
        &StateNode::list("Alpha", [StateNode::singleton("Beta")]),
        (SourceState::from(("GameState", "Alpha")), NamingScheme::Full).into()), @r"
        #[derive(bevy::prelude::SubStates, Hash, Default, Debug, Clone, PartialEq, Eq)]
        #[source(GameStateAlpha = GameStateAlpha::Beta)]
        pub struct GameStateAlphaBeta;
        ");
}

#[cfg(feature = "lists")]
#[rstest]
fn snapshot4() {
    assert_snapshot!(generate_all_type_definitions(
        &StateNode::list("List", [
            StateNode::singleton("Item1"),
            StateNode::enumeration("Item2", [
                StateNode::singleton("A"),
                StateNode::singleton("B"),
            ]),
            StateNode::singleton("Item3"),
        ]),
        (SourceState::from(("GameState", "Alpha")), NamingScheme::Full).into()), @r"
        #[derive(bevy::prelude::SubStates, Hash, Default, Debug, Clone, PartialEq, Eq)]
        #[source(GameStateList = GameStateList::Item1)]
        pub struct GameStateListItem1;
        #[derive(bevy::prelude::SubStates, Hash, Default, Debug, Clone, PartialEq, Eq)]
        #[source(GameStateList = GameStateList::Item2)]
        pub enum GameStateListItem2 { #[default] A, B }

        #[derive(bevy::prelude::SubStates, Hash, Default, Debug, Clone, PartialEq, Eq)]
        #[source(GameStateListItem2 = GameStateListItem2::A)]
        pub struct GameStateListItem2A;
        #[derive(bevy::prelude::SubStates, Hash, Default, Debug, Clone, PartialEq, Eq)]
        #[source(GameStateListItem2 = GameStateListItem2::B)]
        pub struct GameStateListItem2B;
        #[derive(bevy::prelude::SubStates, Hash, Default, Debug, Clone, PartialEq, Eq)]
        #[source(GameStateList = GameStateList::Item3)]
        pub struct GameStateListItem3;
        ");
}
