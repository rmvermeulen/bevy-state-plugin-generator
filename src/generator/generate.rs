use std::rc::Rc;

use indoc::formatdoc;
use itertools::Itertools;
use nom::AsChar;

use crate::PluginConfig;
use crate::config::PluginName;
use crate::generator::context::Context;
use crate::generator::state_defs::generate_all_state_definitions;
use crate::models::{DefinedStates, StateNode};
use crate::parsing::{ParseNode, parse_states_text};
use crate::tree::SubTree;

pub(super) const REQUIRED_DERIVES: &[&str] =
    &["Hash", "Default", "Debug", "Clone", "PartialEq", "Eq"];

pub trait ToStringWith {
    fn to_string_indented<S: AsRef<str>>(&self, join: S) -> String;
}

pub fn get_package_info() -> String {
    let pkg = env!("CARGO_PKG_NAME");
    #[cfg(not(test))]
    let version = env!("CARGO_PKG_VERSION");
    #[cfg(test)]
    let version = "[CARGO_PKG_VERSION]";
    format!("{pkg} v{version}")
}

pub fn generate_debug_info(src_path: &str, source: &str) -> String {
    let lines = source.lines().map(|line| format!("// {line}")).join("\n");
    let pkg_info = get_package_info();
    formatdoc! {"
        // generated by {pkg_info}
        // src: {src_path}
        {lines}
    "}
}

pub(crate) fn generate_plugin_source_from_defined_states(
    defined_states: DefinedStates,
    config: PluginConfig,
) -> String {
    let PluginConfig {
        plugin_name,
        root_state_name,
        states_module_name,
        naming_scheme,
        additional_derives,
    } = config;

    let mut context = Context::from(naming_scheme);
    for derive in additional_derives {
        context.derives.push(derive.to_string());
    }

    let type_definitions = generate_all_state_definitions(defined_states, context);
    let definitions_source = type_definitions.to_string_indented("    ");

    let plugin_builder = if let Some(root_state_name) = root_state_name {
        let init_state = format!(".init_state::<{states_module_name}::{root_state_name}>()");
        let sub_states = type_definitions
            .inner()
            .into_iter()
            .skip(1) // skip root
            .map(|typedef| typedef.typename)
            .map(|state_name| format!(".add_sub_state::<{states_module_name}::{state_name}>()"))
            .join("\n            ");
        format!("app{init_state}{sub_states};")
    } else {
        let states = type_definitions
            .inner()
            .into_iter()
            .map(|sdef| {
                if sdef.parent_name.is_some() {
                    let state_name = sdef.typename;
                    format!(".add_sub_state::<{states_module_name}::{state_name}>()")
                } else {
                    let state_name = sdef.typename;
                    format!(".init_state::<{states_module_name}::{state_name}>()")
                }
            })
            .join("\n            ");
        format!("app{states};")
    };

    let plugin_def = match plugin_name {
        PluginName::Struct(plugin_name) => {
            formatdoc! {"
                pub struct {plugin_name};
                impl bevy::app::Plugin for {plugin_name} {{
                    fn build(&self, app: &mut bevy::app::App) {{
                        {plugin_builder}
                    }}
                }}
            "}
        }
        PluginName::Function(plugin_name) => {
            formatdoc! {"
                pub fn {plugin_name}(app: &mut bevy::app::App) {{
                    {plugin_builder}
                }}
            "}
        }
    };

    formatdoc! {"
        #![allow(missing_docs)]
        use bevy::prelude::AppExtStates;
        pub mod {states_module_name} {{
            use bevy::prelude::StateSet;
            {definitions_source}
        }}
        {plugin_def}
    "}
}

#[cfg(feature = "rustfmt")]
fn try_format_source(source: &str) -> std::io::Result<String> {
    duct::cmd!("rustfmt")
        .stdin_bytes(source)
        .stderr_to_stdout()
        .read()
}

pub(crate) fn format_source<S: AsRef<str>>(source: S) -> String {
    let source = source.as_ref();
    #[cfg(feature = "rustfmt")]
    let source = try_format_source(source).unwrap_or_else(|_| source.to_owned());
    #[cfg(not(feature = "rustfmt"))]
    let source = source.to_owned();

    if source.ends_with(|c: char| c.is_newline()) {
        source
    } else {
        source + "\n"
    }
}

pub fn generate_state_plugin_source(
    source: &str,
    plugin_config: PluginConfig,
    src_path: Option<&str>,
) -> Result<String, String> {
    let parse_nodes = parse_states_text(source).map_err(|e| e.to_string())?;
    let defined_states = if let Some(ref root_state_name) = plugin_config.root_state_name {
        let root_state_name = root_state_name.to_string();
        let parse_tree = if parse_nodes.is_empty() {
            ParseNode::singleton(root_state_name.as_str())
        } else {
            ParseNode::enumeration(root_state_name.as_str(), parse_nodes)
        };
        let parse_tree_size = parse_tree.get_tree_size();
        let root_node: Rc<StateNode> = parse_tree
            .try_into_state_node()
            .map(Rc::new)
            .map_err(|e| format!("{e:?}"))?;
        let state_tree_size = root_node.get_tree_size();

        if state_tree_size > parse_tree_size {
            return Err("state-tree exceeds parse-tree!".into());
        }

        DefinedStates::Root(root_node)
    } else {
        DefinedStates::Unrelated(
            parse_nodes
                .into_iter()
                .map(|node| node.try_into_state_node())
                .collect::<Result<Vec<_>, _>>()
                .unwrap(),
        )
    };

    let plugin_source = generate_plugin_source_from_defined_states(defined_states, plugin_config);
    let source = if let Some(src_path) = src_path {
        let debug_info = generate_debug_info(src_path, source);
        [debug_info, plugin_source].join("\n")
    } else {
        plugin_source
    };

    Ok(format_source(source))
}
