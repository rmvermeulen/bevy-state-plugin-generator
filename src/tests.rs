use std::time::Duration;

use insta::assert_snapshot;
use rstest::rstest;

use crate::generate::{format_source, generate_debug_info, generate_state_plugin_source};
use crate::{NamingScheme, PluginConfig, set_snapshot_suffix};

#[rstest]
#[timeout(Duration::from_millis(250))]
#[async_std::test]
async fn test_format_source() {
    let suffix = cfg!(feature = "rustfmt")
        .then_some("_rustfmt")
        .unwrap_or_default();
    set_snapshot_suffix!("formatted{suffix}");
    let formatted = format_source("fn main(){println!(\"Hello, world!\");}");
    assert_snapshot!(formatted);
}

// #[rstest]
// fn test_generate_states_plugin() {
//     let root_state = StateNode::enumeration(
//         "GameState",
//         [
//             StateNode::singleton("Loading"),
//             StateNode::enumeration(
//                 "Ready",
//                 [
//                     StateNode::enumeration(
//                         "Menu",
//                         [
//                             StateNode::singleton("Main"),
//                             StateNode::singleton("Options"),
//                         ],
//                     ),
//                     StateNode::enumeration(
//                         "Game",
//                         [
//                             StateNode::singleton("Playing"),
//                             StateNode::singleton("Paused"),
//                             StateNode::singleton("GameOver"),
//                         ],
//                     ),
//                 ],
//             ),
//         ],
//     );
//     assert_snapshot!(
//         generate_state_plugin_source(root_state.into(), Default::default(), None).unwrap()
//     );
// }

#[rstest]
#[case("root.txt", "RootState")]
#[case("fruits.txt", "Apple Orange { O1 O2 }")]
fn test_generate_debug_info(#[case] src_path: &str, #[case] source: &str) {
    let suffix = cfg!(feature = "rustfmt")
        .then_some("_rustfmt")
        .unwrap_or_default();
    set_snapshot_suffix!("{src_path}{suffix}");
    assert_snapshot!(generate_debug_info(src_path, source));
}

// TODO: fix test
// #[rstest]
// #[case("comments.txt", StateNode::enum_empty("GameState"))]
// #[case("simple.txt", StateNode::enumeration("GameState", [
//     StateNode::singleton("Loading"),
//     StateNode::enumeration("Ready", [
//         StateNode::singleton("Menu"),
//         StateNode::singleton("Game"),
//     ]),
// ]))]
// #[case("fruits.txt", StateNode::enumeration("GameState", [
//     StateNode::singleton("Loading"),
//     StateNode::enumeration("Ready", [
//         StateNode::enumeration("Menu", [
//             StateNode::singleton("Main"),
//             StateNode::singleton("Options"),
//         ]),
//         StateNode::enumeration("Game", [
//             StateNode::singleton("Playing"),
//             StateNode::singleton("Paused"),
//             StateNode::singleton("GameOver"),
//         ]),
//     ]),
// ]))]
// fn test_generate_plugin_source_inner(#[case] src_path: &str, #[case] root_node: StateNode) {
//     let suffix = cfg!(feature = "rustfmt")
//         .then_some("_rustfmt")
//         .unwrap_or_default();
//     set_snapshot_suffix!("{src_path}{suffix}");
// assert_snapshot!(format_source(generate_state_plugin_source(
//     root_node.into(),
//     Default::default(),
//     None
// )));
// }

#[rstest]
#[case("root.txt", "RootState", PluginConfig::default())]
#[case(
    "mixed-nested-states.txt",
    "A { B [C[ } D { E F [ G H ] I }",
    PluginConfig::default()
)]
fn test_generate_full_source(
    #[case] src_path: &str,
    #[case] source: &str,
    #[case] plugin_config: PluginConfig,
) {
    let suffix = cfg!(feature = "rustfmt")
        .then_some("_rustfmt")
        .unwrap_or_default();
    set_snapshot_suffix!("{src_path}{suffix}");
    assert_snapshot!(generate_state_plugin_source(source, plugin_config, Some(src_path)).unwrap());
}

#[rstest]
#[case("root.txt", "RootState", NamingScheme::Full)]
#[case("root.txt", "RootState", NamingScheme::Short)]
fn test_naming_scheme(
    #[case] src_path: &str,
    #[case] source: &str,
    #[case] naming_scheme: NamingScheme,
) {
    use crate::generate::generate_state_plugin_source;

    let suffix = cfg!(feature = "rustfmt")
        .then_some("_rustfmt")
        .unwrap_or_default();
    set_snapshot_suffix!("{src_path}_{naming_scheme}{suffix}");
    assert_snapshot!(
        generate_state_plugin_source(
            source,
            PluginConfig {
                naming_scheme,
                ..Default::default()
            },
            Some(src_path),
        )
        .unwrap()
    );
}
