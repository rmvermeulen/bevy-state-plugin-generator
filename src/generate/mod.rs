pub(crate) mod core;
#[cfg(test)]
mod tests;

use std::path::Path;
use std::{fs, io};

use itertools::{Itertools, concat};

use crate::generate::core::generate_state_plugin_source;
use crate::parsing::header::parse_template_header;
use crate::prelude::PluginConfig;
use crate::processing::ProcessingError;

/// The kinds of errors that can occur
#[derive(Debug, thiserror::Error)]
pub enum GeneratorError {
    /// A fs-related error occurred
    #[error("Io Error: {0}")]
    Io(#[from] io::Error),
    /// The content is wrong
    #[error("Processing Error: {0}")]
    Processing(#[from] ProcessingError),
}

/// ```rust no_run
/// use bevy_state_plugin_generator::prelude::*;
/// fn main() {
///   update_template(
///     "src/generated_states.rs",
///     PluginConfig::default()
///   ).expect("Failed to generate plugin!");
/// }
/// ```
#[cfg_attr(coverage_nightly, coverage(off))]
pub fn update_template(
    template_path: impl AsRef<Path>,
    mut plugin_config: PluginConfig,
) -> Result<(), GeneratorError> {
    let src_display = template_path.as_ref().to_string_lossy();
    println!("cargo:rerun-if-changed={src_display}");
    let source = std::fs::read_to_string(&template_path)?;
    let header = parse_template_header(&source, &mut plugin_config);

    let plugin_source =
        generate_state_plugin_source(&header.template.join("\n"), plugin_config, None)?;

    let header = concat([
        header
            .info_block
            .into_iter()
            .map(|line| format!("// {line}"))
            .collect_vec(),
        // vec![
        //     format!("comment_block={comment_block:?}"),
        //     format!("info_and_warnings={info_and_warnings:?}"),
        // ],
        header
            .comments_block
            .into_iter()
            .map(String::from)
            .collect_vec(),
    ])
    .join("\n");

    fs::write(&template_path, format!("{header}\n\n{plugin_source}")).map_err(Into::into)
}

/// ```rust no_run
/// use bevy_state_plugin_generator::prelude::*;
/// fn main() {
///   generate_plugin(
///     "src/states.txt",
///     "src/generated_states.rs",
///     PluginConfig::default()
///   ).expect("Failed to generate plugin!");
/// }
/// ```
#[cfg_attr(coverage_nightly, coverage(off))]
pub fn generate_plugin(
    src: impl AsRef<Path>,
    dst: impl AsRef<Path>,
    plugin_config: PluginConfig,
) -> Result<(), GeneratorError> {
    let src_display = src.as_ref().to_string_lossy();
    println!("cargo:rerun-if-changed={src_display}");
    let source = std::fs::read_to_string(&src)?;
    let source = generate_state_plugin_source(&source, plugin_config, Some(&src_display))?;
    fs::write(dst, source).map_err(Into::into)
}
