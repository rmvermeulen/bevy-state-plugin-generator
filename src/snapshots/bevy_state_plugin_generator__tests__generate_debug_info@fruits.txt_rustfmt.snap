---
source: src/tests.rs
expression: "generate_debug_info(src_path, source)"
---
// generated by bevy_state_plugin_generator v[CARGO_PKG_VERSION]
// src: fruits.txt
// Apple Orange { O1 O2 }
